core/drivers/firewall/stm32_risaf.c

```diff
@@ -0,0 +1,770 @@
+#define _RISAF_GET_REGION_CFG(cfg) \
+	(((((cfg) & DT_RISAF_EN_MASK) >> DT_RISAF_EN_SHIFT) << \
+	  _RISAF_REG_CFGR_BREN_SHIFT) | \
+	 ((((cfg) & DT_RISAF_SEC_MASK) >> DT_RISAF_SEC_SHIFT) << \
+	  _RISAF_REG_CFGR_SEC_SHIFT) | \
+	 ((((cfg) & DT_RISAF_ENC_MASK) >> (DT_RISAF_ENC_SHIFT + 1)) << \
+	  _RISAF_REG_CFGR_ENC_SHIFT) | \
+	 ((((cfg) & DT_RISAF_PRIV_MASK) >> DT_RISAF_PRIV_SHIFT) << \
+	  _RISAF_REG_CFGR_PRIVC_SHIFT))
+#define _RISAF_GET_REGION_CID_CFG(cfg) \
+	(((((cfg) & DT_RISAF_WRITE_MASK) >> DT_RISAF_WRITE_SHIFT) << \
+	  _RISAF_REG_CIDCFGR_WRENC_SHIFT) | \
+	 ((((cfg) & DT_RISAF_READ_MASK) >> DT_RISAF_READ_SHIFT) << \
+	  _RISAF_REG_CIDCFGR_RDENC_SHIFT))
+
+#ifdef CFG_CORE_RESERVED_SHM
+#define TZDRAM_RESERVED_SIZE		(TZDRAM_SIZE + TEE_SHMEM_SIZE)
+#else
+#define TZDRAM_RESERVED_SIZE		TZDRAM_SIZE
+#endif
+
+#define _RISAF_NB_CID_SUPPORTED		U(8)
+
+struct stm32_risaf_region {
+	uint32_t cfg;
+	uintptr_t addr;
+	size_t len;
+};
+
+struct stm32_risaf_pdata {
+	struct io_pa_va base;
+	struct clk *clock;
+	struct stm32_risaf_region *regions;
+	char risaf_name[20];
+	unsigned int nregions;
+	unsigned int conf_lock;
+	uintptr_t mem_base;
+	size_t mem_size;
+	bool enc_supported;
+};
+
+struct stm32_risaf_ddata {
+	uint32_t mask_regions;
+	uint32_t max_base_regions;
+	uint32_t granularity;
+};
+
+struct stm32_risaf_instance {
+	struct stm32_risaf_pdata pdata;
+	struct stm32_risaf_ddata *ddata;
+
+	SLIST_ENTRY(stm32_risaf_instance) link;
+};
+
+struct stm32_risaf_version {
+	uint32_t major;
+	uint32_t minor;
+	uint32_t ip_id;
+	uint32_t size_id;
+};
+
+/**
+ * struct stm32_risaf_compat_data describes RISAF associated data
+ * for compatible list.
+ *
+ * @supported_encryption:	identify risaf encryption capabilities.
+ */
+struct stm32_risaf_compat_data {
+	bool supported_encryption;
+};
+
+static bool is_tdcid;
+
+static const struct stm32_risaf_compat_data stm32_risaf_compat = {
+	.supported_encryption = false,
+};
+
+static const struct stm32_risaf_compat_data stm32_risaf_enc_compat = {
+	.supported_encryption = true,
+};
+
+static SLIST_HEAD(, stm32_risaf_instance) risaf_list =
+		SLIST_HEAD_INITIALIZER(risaf_list);
+
+static vaddr_t risaf_base(struct stm32_risaf_instance *risaf)
+{
+	return io_pa_or_va_secure(&risaf->pdata.base, 1);
+}
+
+void stm32_risaf_clear_illegal_access_flags(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		io_write32(base + _RISAF_IACR, _RISAF_IACR_CAEF |
+			   _RISAF_IACR_IAEF0 | _RISAF_IACR_IAEF1);
+
+		clk_disable(risaf->pdata.clock);
+	}
+}
+
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risaf_dump_erroneous_data(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
```

Comment:
What about the below, to simplify the header file:
```c
	if (!IS_ENABLED(CFG_TEE_CORE_DEBUG))
		reutrn;

Reply:
mmm okay, it involves a bigger footprint but I hope it can be optimized

Reply:
I think modern toolchains (used with `-Os` or `-O2`) will discard the unused code/data.

Reply:
Actually, I prefer to keep current implementation if that's ok

Reply:
Ok, let's implement it like that then